File: src/main/java/com/example/counter/challenge/ChallengeService.java
package com.example.counter.challenge;

import com.example.counter.user.User;
import com.example.counter.user.UserRepository;
import com.example.counter.habit.Habit;
import com.example.counter.habit.HabitEntry;
import com.example.counter.habit.HabitRepository;
import com.example.counter.habit.HabitEntryRepository;
import com.example.counter.challenge.ChallengeDto.*;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StopWatch;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.List;
import java.util.Optional;

@Service
public class ChallengeService {
    private final ChallengeRepository challengeRepo;
    private final ChallengeHabitsRepository challengeHabitsRepo;
    private final HabitRepository habitRepo;
    private final HabitEntryRepository habitEntryRepo;
    private final UserRepository userRepo;

    public ChallengeService(ChallengeRepository challengeRepo, 
                          ChallengeHabitsRepository challengeHabitsRepo,
                          HabitRepository habitRepo,
                          HabitEntryRepository habitEntryRepo,
                          UserRepository userRepo) {
        this.challengeRepo = challengeRepo;
        this.challengeHabitsRepo = challengeHabitsRepo;
        this.habitRepo = habitRepo;
        this.habitEntryRepo = habitEntryRepo;
        this.userRepo = userRepo;
    }

    @Transactional(rollbackFor = Exception.class)
    public ChallengeResponse createChallenge(String username, ChallengeCreateRequest request) {
        return executeWithTiming("create challenge", () -> {
            System.out.println("createChallengeRequest: " + request.toString());

            User user = validateAndGetUser(username);
            validateChallengeUniqueness(username, request.name(), request.startDate());
            List<Habit> habits = validateAndGetHabits(username, request.habitIds());

            // Create challenge
            Challenge challenge = new Challenge(user, request.name(), request.startDate(), request.durationDays());
            challenge = challengeRepo.save(challenge);

            // Add habits to challenge
            for (Habit habit : habits) {
                ChallengeHabits challengeHabit = new ChallengeHabits(challenge, habit);
                challengeHabitsRepo.save(challengeHabit);
            }

            return new ChallengeResponse(challenge);
        });
    }

    @Transactional
    public List<ChallengeSummaryResponse> getChallengesByStatus(String username, ChallengeStatus status) {
        return executeWithTiming("get challenges by status", () -> {
            List<Challenge> challenges;
            if (status != null) {
                challenges = challengeRepo.findByUsernameAndStatus(username, status);
            } else {
                challenges = challengeRepo.findByUsernameAndNotDeleted(username);
            }

            // Update status of scheduled and active challenges based on current date
            updateChallengeStatuses(challenges);

            return challenges.stream()
                    .map(ChallengeSummaryResponse::new)
                    .toList();
        });
    }

    public ChallengeDetailResponse getChallengeDetails(String username, Long challengeId) {
        return executeWithTiming("get challenge details", () -> {
            Challenge challenge = validateAndGetChallenge(username, challengeId);
            List<Habit> habits = challengeHabitsRepo.findHabitsByChallengeId(challengeId);
            List<HabitInfo> habitsInfo = processHabitEntriesForChallenge(challenge, habits);
            
            return new ChallengeDetailResponse(challenge, habitsInfo);
        });
    }

    @Transactional
    public ChallengeResponse updateChallenge(String username, Long challengeId, ChallengeUpdateRequest request) {
        return executeWithTiming("update challenge", () -> {
            Challenge challenge = validateAndGetChallenge(username, challengeId);

            boolean updated = false;

            if (request.name() != null && !request.name().trim().isEmpty()) {
                challenge.setName(request.name());
                updated = true;
            }

            if (request.startDate() != null) {
                challenge.setStartDate(request.startDate());
                updated = true;
            }

            if (request.durationDays() != null) {
                challenge.setDurationDays(request.durationDays());
                updated = true;
            }

            if (updated) {
                challenge = challengeRepo.save(challenge);
            }

            return new ChallengeResponse(challenge);
        });
    }

    @Transactional
    public void deleteChallenge(String username, Long challengeId) {
        executeWithTiming("delete challenge", () -> {
            Challenge challenge = validateAndGetChallenge(username, challengeId);
            // Soft delete by setting status to deleted
            challenge.setScheduleStatus(ChallengeStatus.deleted);
            challengeRepo.save(challenge);
        });
    }

    @Transactional
    public void addHabitToChallenge(String username, Long challengeId, Long habitId) {
        executeWithTiming("add habit to challenge", () -> {
            Challenge challenge = validateAndGetChallenge(username, challengeId);
            Habit habit = validateAndGetHabit(username, habitId);

            // Check if habit is already in challenge
            Optional<ChallengeHabits> existing = challengeHabitsRepo.findByChallengeIdAndHabitId(challengeId, habitId);
            if (existing.isPresent()) {
                throw new RuntimeException("Habit is already added to this challenge");
            }

            // Add habit to challenge
            ChallengeHabits challengeHabit = new ChallengeHabits(challenge, habit);
            challengeHabitsRepo.save(challengeHabit);
        });
    }

    @Transactional
    public void deleteHabitFromChallenge(String username, Long challengeId, Long habitId) {
        executeWithTiming("delete habit from challenge", () -> {
            validateAndGetChallenge(username, challengeId);
            validateAndGetHabit(username, habitId);

            // Check if habit is in challenge
            Optional<ChallengeHabits> existing = challengeHabitsRepo.findByChallengeIdAndHabitId(challengeId, habitId);
            if (existing.isEmpty()) {
                throw new RuntimeException("Habit is not in this challenge");
            }

            // Remove habit from challenge
            challengeHabitsRepo.deleteByChallengeIdAndHabitId(challengeId, habitId);
        });
    }

    // ==================== HELPER METHODS ====================
    
    /**
     * Validates that a user exists and returns the user entity
     */
    private User validateAndGetUser(String username) {
        return userRepo.findByUsername(username)
            .orElseThrow(() -> new RuntimeException("User not found: " + username));
    }
    
    /**
     * Validates that a challenge exists and belongs to the user, returns the challenge entity
     */
    private Challenge validateAndGetChallenge(String username, Long challengeId) {
        return challengeRepo.findByUsernameAndChallengeIdAndNotDeleted(username, challengeId)
            .orElseThrow(() -> new RuntimeException("Challenge not found with id: " + challengeId));
    }
    
    /**
     * Validates that a habit exists and belongs to the user, returns the habit entity
     */
    private Habit validateAndGetHabit(String username, Long habitId) {
        Habit habit = habitRepo.findById(habitId)
            .orElseThrow(() -> new RuntimeException("Habit not found with id: " + habitId));
        
        if (!habit.getUser().getUsername().equals(username)) {
            throw new RuntimeException("Habit does not belong to user: " + username);
        }
        
        return habit;
    }
    
    /**
     * Validates that multiple habits exist and belong to the user
     */
    private List<Habit> validateAndGetHabits(String username, List<Long> habitIds) {
        List<Habit> habits = habitRepo.findHabitsByIdsAndUsername(habitIds, username);
        if (habits.size() != habitIds.size()) {
            throw new RuntimeException("One or more habit IDs do not exist or do not belong to user: " + username);
        }
        return habits;
    }
    
    /**
     * Checks if a challenge with the same name and start date already exists
     */
    private void validateChallengeUniqueness(String username, String name, java.time.LocalDate startDate) {
        Optional<Challenge> existingChallenge = challengeRepo.findByUsernameAndNameAndStartDateAndNotDeleted(
            username, name, startDate);
        if (existingChallenge.isPresent()) {
            throw new RuntimeException("Challenge with name '" + name + 
                "' and start date '" + startDate + "' already exists for user: " + username);
        }
    }
    
    /**
     * Processes habit entries for a challenge, filtering by date range and grouping by habit
     */
    private List<HabitInfo> processHabitEntriesForChallenge(Challenge challenge, List<Habit> habits) {
        // Get all habit entries for these habits within the challenge date range
        List<Long> habitIds = habits.stream().map(Habit::getHabitId).toList();
        List<HabitEntry> allEntries = habitEntryRepo.findByHabitIds(habitIds);
        
        // Filter entries within challenge date range
        List<HabitEntry> challengeEntries = allEntries.stream()
            .filter(entry -> !entry.getEntryDate().isBefore(challenge.getStartDate()) && 
                           !entry.getEntryDate().isAfter(challenge.getEndDate()))
            .toList();

        // Group entries by habit
        return habits.stream()
            .map(habit -> {
                List<HabitEntry> habitEntries = challengeEntries.stream()
                    .filter(entry -> entry.getHabit().getHabitId().equals(habit.getHabitId()))
                    .toList();
                return new HabitInfo(habit, habitEntries);
            })
            .toList();
    }
    
    /**
     * Wrapper for performance logging - executes a function and logs the time taken
     */
    private <T> T executeWithTiming(String operationName, java.util.function.Supplier<T> operation) {
        StopWatch stopWatch = new StopWatch();
        stopWatch.start();
        
        try {
            return operation.get();
        } finally {
            stopWatch.stop();
            logTime(operationName + " took ms = ", stopWatch);
        }
    }
    
    /**
     * Wrapper for performance logging - executes a void function and logs the time taken
     */
    private void executeWithTiming(String operationName, Runnable operation) {
        executeWithTiming(operationName, () -> {
            operation.run();
            return null;
        });
    }
    
    /**
     * Updates challenge statuses based on current date in India timezone
     * Only updates scheduled and active challenges that need status changes
     */
    private void updateChallengeStatuses(List<Challenge> challenges) {
        // Get current date in India timezone
        ZoneId indiaZone = ZoneId.of("Asia/Kolkata");
        LocalDate today = LocalDate.now(indiaZone);
        
        boolean hasUpdates = false;
        
        for (Challenge challenge : challenges) {
            // Only update scheduled and active challenges
            if (challenge.getScheduleStatus() == ChallengeStatus.scheduled || 
                challenge.getScheduleStatus() == ChallengeStatus.active) {
                
                ChallengeStatus newStatus = determineChallengeStatus(challenge, today);
                
                // Only update if status has changed
                if (challenge.getScheduleStatus() != newStatus) {
                    challenge.setScheduleStatus(newStatus);
                    hasUpdates = true;
                    System.out.println("Updated challenge '" + challenge.getName() + 
                        "' status from " + challenge.getScheduleStatus() + " to " + newStatus);
                }
            }
        }
        
        // Save all updated challenges in one batch
        if (hasUpdates) {
            challengeRepo.saveAll(challenges);
        }
    }
    
    /**
     * Determines the appropriate status for a challenge based on current date
     */
    private ChallengeStatus determineChallengeStatus(Challenge challenge, LocalDate today) {
        LocalDate startDate = challenge.getStartDate();
        LocalDate endDate = challenge.getEndDate();
        
        if (startDate.isAfter(today)) {
            return ChallengeStatus.scheduled;
        } else if (startDate.isEqual(today) || (startDate.isBefore(today) && endDate.isAfter(today)) || endDate.isEqual(today)) {
            return ChallengeStatus.active;
        } else {
            return ChallengeStatus.expired;
        }
    }

    /**
     * Logs the time taken for an operation
     */
    private void logTime(String message, StopWatch stopWatch) {
        System.out.println(message + stopWatch.getTotalTimeMillis() + " ms");
    }
}

File: src/main/kotlin/com/example/counter/challenge/ChallengeService.kt
package com.example.counter.challenge

import com.example.counter.user.User
import com.example.counter.user.UserRepository
import com.example.counter.habit.Habit
import com.example.counter.habit.HabitEntry
import com.example.counter.habit.HabitRepository
import com.example.counter.habit.HabitEntryRepository
import com.example.counter.notes.Note
import com.example.counter.notes.NoteRepository
import com.example.counter.notes.NoteDto.NoteResponse
import com.example.counter.challenge.ChallengeDto.*
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.StopWatch
import java.time.LocalDate
import java.time.ZoneId
import java.util.Optional
import java.util.function.Supplier

@Service
class ChallengeService(
    private val challengeRepo: ChallengeRepository,
    private val challengeHabitsRepo: ChallengeHabitsRepository,
    private val habitRepo: HabitRepository,
    private val habitEntryRepo: HabitEntryRepository,
    private val userRepo: UserRepository,
    private val noteRepo: NoteRepository
) {

    @Transactional(rollbackFor = [Exception::class])
    fun createChallenge(username: String, request: ChallengeCreateRequest): ChallengeResponse {
        return executeWithTiming("create challenge") {
            println("createChallengeRequest: $request")

            val user = validateAndGetUser(username)
            validateChallengeUniqueness(username, request.name, request.startDate)
            val habits = validateAndGetHabits(username, request.habitIds)

            // Create challenge
            var challenge = Challenge(user, request.name, request.startDate, request.durationDays)
            challenge = challengeRepo.save(challenge)

            // Add habits to challenge
            for (habit in habits) {
                val challengeHabit = ChallengeHabits(challenge, habit)
                challengeHabitsRepo.save(challengeHabit)
            }

            ChallengeResponse(challenge)
        }
    }

    @Transactional
    fun getChallengesByStatus(username: String, status: ChallengeStatus?): List<ChallengeSummaryResponse> {
        return executeWithTiming("get challenges by status") {
            val challenges = if (status != null) {
                challengeRepo.findByUsernameAndStatus(username, status)
            } else {
                challengeRepo.findByUsernameAndNotDeleted(username)
            }

            // Update status of scheduled and active challenges based on current date
            updateChallengeStatuses(challenges)

            challenges.map { ChallengeSummaryResponse(it) }
        }
    }

    fun getChallengeDetails(username: String, challengeId: Long): ChallengeDetailResponse {
        return executeWithTiming("get challenge details") {
            val challenge = validateAndGetChallenge(username, challengeId)
            val habits = challengeHabitsRepo.findHabitsByChallengeId(challengeId)
            val habitsInfo = processHabitEntriesForChallenge(challenge, habits)
            
            ChallengeDetailResponse(challenge, habitsInfo)
        }
    }

    @Transactional
    fun updateChallenge(username: String, challengeId: Long, request: ChallengeUpdateRequest): ChallengeResponse {
        return executeWithTiming("update challenge") {
            var challenge = validateAndGetChallenge(username, challengeId)
            var updated = false

            if (!request.name.isNullOrBlank()) {
                challenge.name = request.name
                updated = true
            }

            if (request.startDate != null) {
                challenge.startDate = request.startDate
                updated = true
            }

            if (request.durationDays != null) {
                challenge.durationDays = request.durationDays
                updated = true
            }

            if (updated) {
                challenge = challengeRepo.save(challenge)
            }

            ChallengeResponse(challenge)
        }
    }

    @Transactional
    fun deleteChallenge(username: String, challengeId: Long) {
        executeWithTiming("delete challenge") {
            val challenge = validateAndGetChallenge(username, challengeId)
            // Soft delete by setting status to deleted
            challenge.scheduleStatus = ChallengeStatus.deleted
            challengeRepo.save(challenge)
        }
    }

    @Transactional
    fun addHabitToChallenge(username: String, challengeId: Long, habitId: Long) {
        executeWithTiming("add habit to challenge") {
            val challenge = validateAndGetChallenge(username, challengeId)
            val habit = validateAndGetHabit(username, habitId)

            // Check if habit is already in challenge
            val existing = challengeHabitsRepo.findByChallengeIdAndHabitId(challengeId, habitId)
            if (existing != null) {
                throw RuntimeException("Habit is already added to this challenge")
            }

            // Add habit to challenge
            val challengeHabit = ChallengeHabits(challenge, habit)
            challengeHabitsRepo.save(challengeHabit)
        }
    }

    @Transactional
    fun deleteHabitFromChallenge(username: String, challengeId: Long, habitId: Long) {
        executeWithTiming("delete habit from challenge") {
            validateAndGetChallenge(username, challengeId)
            validateAndGetHabit(username, habitId)

            // Check if habit is in challenge
            val existing = challengeHabitsRepo.findByChallengeIdAndHabitId(challengeId, habitId)
            if (existing == null) {
                throw RuntimeException("Habit is not in this challenge")
            }

            // Remove habit from challenge
            challengeHabitsRepo.deleteByChallengeIdAndHabitId(challengeId, habitId)
        }
    }

    @Transactional
    fun upsertNoteToChallenge(username: String, challengeId: Long, noteDate: LocalDate, noteText: String): NoteResponse {
        return executeWithTiming("upsert note to challenge") {
            val challenge = validateAndGetChallenge(username, challengeId)
            val user = validateAndGetUser(username)
            
            // Check if note exists for this date without challenge
            val existingNote = noteRepo.findByUserAndNoteDate(user, noteDate)
            if (existingNote.isPresent) {
                updateExistingNote(existingNote.get(), noteText, challenge)
            } else {
                createNewNote(user, noteText, challenge, noteDate)
            }
        }
    }

    // ==================== HELPER METHODS ====================
    
    private fun updateExistingNote(note: Note, noteText: String, challenge: Challenge): NoteResponse {
        val message: String
        if (note.challenge == null) {
            note.challenge = challenge
            message = "Note already existed and has been attached to the challenge"
        } else {
            note.noteText = noteText
            message = "Note updated successfully"
        }
        noteRepo.save(note)
        return NoteResponse(note, message)
    }

    private fun createNewNote(user: User, noteText: String, challenge: Challenge, noteDate: LocalDate): NoteResponse {
        var newNote = Note(user, noteText, noteDate, challenge)
        newNote = noteRepo.save(newNote)
        return NoteResponse(newNote, "Note created successfully")
    }

    private fun validateAndGetUser(username: String): User {
        return userRepo.findByUsername(username)
            .orElseThrow { RuntimeException("User not found: $username") }
    }
    
    private fun validateAndGetChallenge(username: String, challengeId: Long): Challenge {
        return challengeRepo.findByUsernameAndChallengeIdAndNotDeleted(username, challengeId)
            .orElseThrow { RuntimeException("Challenge not found with id: $challengeId") }
    }
    
    private fun validateAndGetHabit(username: String, habitId: Long): Habit {
        val habit = habitRepo.findById(habitId)
            .orElseThrow { RuntimeException("Habit not found with id: $habitId") }
        
        if (habit.user.username != username) {
            throw RuntimeException("Habit does not belong to user: $username")
        }
        
        return habit
    }
    
    private fun validateAndGetHabits(username: String, habitIds: List<Long>): List<Habit> {
        val habits = habitRepo.findHabitsByIdsAndUsername(habitIds, username)
        if (habits.size != habitIds.size) {
            throw RuntimeException("One or more habit IDs do not exist or do not belong to user: $username")
        }
        return habits
    }
    
    private fun validateChallengeUniqueness(username: String, name: String, startDate: LocalDate) {
        val existingChallenge = challengeRepo.findByUsernameAndNameAndStartDateAndNotDeleted(
            username, name, startDate
        )
        if (existingChallenge.isPresent) {
            throw RuntimeException("Challenge with name '$name' and start date '$startDate' already exists for user: $username")
        }
    }
    
    private fun processHabitEntriesForChallenge(challenge: Challenge, habits: List<Habit>): List<HabitInfo> {
        val habitIds = habits.mapNotNull { it.habitId }
        val allEntries = habitEntryRepo.findByHabitIds(habitIds)
        
        val challengeEntries = allEntries.filter { entry ->
            !entry.entryDate.isBefore(challenge.startDate) && 
            (challenge.endDate == null || !entry.entryDate.isAfter(challenge.endDate))
        }

        return habits.map { habit ->
            val habitEntries = challengeEntries.filter { entry ->
                entry.habit.habitId == habit.habitId
            }
            HabitInfo(habit, habitEntries)
        }
    }
    
    private fun <T> executeWithTiming(operationName: String, operation: () -> T): T {
        val stopWatch = StopWatch()
        stopWatch.start()
        
        try {
            return operation()
        } finally {
            stopWatch.stop()
            logTime("$operationName took ms = ", stopWatch)
        }
    }
    
    private fun updateChallengeStatuses(challenges: List<Challenge>) {
        val indiaZone = ZoneId.of("Asia/Kolkata")
        val today = LocalDate.now(indiaZone)
        
        var hasUpdates = false
        
        for (challenge in challenges) {
            if (challenge.scheduleStatus == ChallengeStatus.scheduled || 
                challenge.scheduleStatus == ChallengeStatus.active) {
                
                val newStatus = determineChallengeStatus(challenge, today)
                
                if (challenge.scheduleStatus != newStatus) {
                    challenge.scheduleStatus = newStatus
                    hasUpdates = true
                    println("Updated challenge '${challenge.name}' status from ${challenge.scheduleStatus} to $newStatus")
                }
            }
        }
        
        if (hasUpdates) {
            challengeRepo.saveAll(challenges)
        }
    }
    
    private fun determineChallengeStatus(challenge: Challenge, today: LocalDate): ChallengeStatus {
        val startDate = challenge.startDate
        val endDate = challenge.endDate
        
        return if (startDate.isAfter(today)) {
            ChallengeStatus.scheduled
        } else if (startDate.isEqual(today) || (startDate.isBefore(today) && (endDate == null || endDate.isAfter(today))) || endDate?.isEqual(today) == true) {
            ChallengeStatus.active
        } else {
            ChallengeStatus.expired
        }
    }

    private fun logTime(message: String, stopWatch: StopWatch) {
        println(message + stopWatch.totalTimeMillis + " ms")
    }
}
